varats/varats/plots/bug_fixing_relations.py

```diff
@@ -0,0 +1,340 @@
+    interval indices are in [0,3] for 5 thresholds."""
+    k = 0
+    while __distance_thresholds[k] < distance:
+        k += 1
+    return k - 1
+
+
+#defining some constants for diagram generation
+__cp_parameters = [1.2, 1.5, 1.8, 2.1]
+__distance_thresholds = [
+    0,
+    _get_distance([1, 0], 2 * [np.sqrt(2) / 2]),
+    np.sqrt(2),
+    _get_distance([1, 0], [-np.sqrt(2) / 2, np.sqrt(2) / 2]), 2.0
+]
+
+
+def _get_bezier_curve(ctrl_points: np.array, num_points: int = 5) -> np.array:
+    """Implements bezier edges to display between commit nodes."""
+    n = len(ctrl_points)
+
+    def get_coordinate_on_curve(factor: float) -> np.array:
+        points_cp = np.copy(ctrl_points)
+        for r in range(1, n):
+            points_cp[:n - r, :] = (
+                1 - factor
+            ) * points_cp[:n - r, :] + factor * points_cp[1:n - r + 1, :]
+        return np.array(points_cp[0, :])
+
+    point_space = np.linspace(0, 1, num_points)
+    return np.array([
+        get_coordinate_on_curve(point_space[k]) for k in range(num_points)
+    ])
+
+
+def _compute_node_placement(commit_count: int) -> tp.List[np.array]:
+    """Compute unit circle coordinates for each commit; move unit circle such
+    that HEAD is on top."""
+    # use commit_count + 1 since first and last coordinates are equal
+    theta_vals = np.linspace(-3 * np.pi / 2, np.pi / 2, commit_count + 1)
+    commit_coordinates: tp.List[np.array] = list()
+    for theta in theta_vals:
+        commit_coordinates.append(np.array([np.cos(theta), np.sin(theta)]))
+    return commit_coordinates
+
+
+def _map_commits_to_nodes(project_repo: pygit2.Repository) -> tp.Dict[str, int]:
+    """Maps commit hex -> node id."""
+    commits_to_nodes_map: tp.Dict[str, int] = {}
+    commit_count = 0
+    for commit in project_repo.walk(
+        project_repo.head.target.hex, pygit2.GIT_SORT_TIME
+    ):
+        # node ids are sorted by time
+        commits_to_nodes_map[commit.hex] = commit_count
+        commit_count += 1
+    return commits_to_nodes_map
+
+
+class BugFixingRelationPlot(Plot):
+    """Plot showing which commit fixed a bug introduced by which commit."""
+
+    NAME = 'bug_relation_graph'
+
+    def __init__(self, **kwargs: tp.Any) -> None:
+        super().__init__(self.NAME, **kwargs)
+
+    @staticmethod
+    def supports_stage_separation() -> bool:
+        return False
+
+    def plot(self, view_mode: bool) -> None:
+        """Plots bug plot for the whole project."""
+        project_name = self.plot_kwargs['project']
+
+        self.__szz_tool = self.plot_kwargs.get('szz_tool', 'provider')
+
+        if self.__szz_tool == 'provider':
+            bug_provider = BugProvider.get_provider_for_project(
+                get_project_cls_by_name(project_name)
+            )
+            raw_bugs = bug_provider.find_all_raw_bugs()
+        elif self.__szz_tool == 'szz_unleashed':
+            pass
+        else:
+            raise PlotDataEmpty
```

Comment:
```suggestion
        elif self.__szz_tool != 'szz_unleashed':
            raise PlotDataEmpty
```
Would this also work? Would make it a bit clearer when the expection is raised.

Furthermore, I'm not sure if `PlotDataEmpty` is the correct exception here, maybe we should build a custom one like: `MissingSZZTool`?
@boehmseb Do you have an opinion about the exception/exception name?

Reply:
This code does make sense as it is. This section retrieves the data depending on the selected `szz_tool`. The validation as you suggest should be already handled in the driver.
The only thing missing here is to replace the `pass` with the actual method to retrieve data from an `SZZReport`.

Reply:
This is implemented now, the `szz_tool` can either be `pydriller`, `szz_unleashed` or `szz_diff`, which creates a different plot each.